File: frontend/lib/core/models/task.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'task.freezed.dart';
part 'task.g.dart';

enum TaskPriority {
  @JsonValue(1)
  low,
  @JsonValue(2) 
  medium,
  @JsonValue(3)
  high,
}

@freezed
class Task with _$Task {
  const factory Task({
    @JsonKey(name: '_id') required String id,
    required String userId,
    required String title,
    String? description,
    @Default(false) bool isCompleted,
    @Default(TaskPriority.medium) TaskPriority priority,
    @JsonKey(name: 'dueAt') DateTime? dueDate,
    String? listId,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = _Task;

  factory Task.fromJson(Map<String, dynamic> json) => _$TaskFromJson(json);
}

File: frontend/src/lib/types/task.ts
// Task type definitions

export type TaskPriority = 1 | 2 | 3; // 1=low, 2=normal, 3=high

export type NotificationTiming = 'same_day' | '1_day_before' | '2_days_before';

export interface ChecklistItem {
  _id?: string;
  text: string;
  isCompleted: boolean;
  order: number;
}

export interface Task {
  _id: string;
  userId: string;
  listId?: string;
  title: string;
  description: string;
  dueAt?: string;
  isCompleted: boolean;
  priority: TaskPriority;
  checklistItems: ChecklistItem[];
  notificationEnabled: boolean;
  notificationTimings: NotificationTiming[];
  notificationsSent?: NotificationTiming[];
  createdAt: string;
  updatedAt: string;
}

export interface CreateTaskData {
  title: string;
  description: string;
  dueAt?: string;
  priority?: TaskPriority;
  listId?: string;
  checklistItems?: ChecklistItem[];
  notificationEnabled?: boolean;
  notificationTimings?: NotificationTiming[];
}

export interface UpdateTaskData extends Partial<CreateTaskData> {
  isCompleted?: boolean;
  checklistItems?: ChecklistItem[];
  notificationEnabled?: boolean;
  notificationTimings?: NotificationTiming[];
}

export interface TaskFilters {
  listId?: string;
  isCompleted?: boolean;
  priority?: TaskPriority;
}
